File: rust/ampersend-sdk/tests/wallet_test.rs
use std::collections::HashMap;

use ampersend_sdk::x402::types::PaymentRequirements;
use ampersend_sdk::x402::wallet::X402Wallet;
use ampersend_sdk::x402::wallets::{AccountWallet, SmartAccountConfig, SmartAccountWallet};

fn test_requirements() -> PaymentRequirements {
    let mut extra = HashMap::new();
    extra.insert("name".to_string(), serde_json::json!("USD Coin"));
    extra.insert("version".to_string(), serde_json::json!("2"));

    PaymentRequirements {
        scheme: "exact".to_string(),
        network: "base-sepolia".to_string(),
        max_amount_required: "10000".to_string(),
        resource: "https://example.com/api".to_string(),
        description: "Test payment".to_string(),
        mime_type: "application/json".to_string(),
        pay_to: "0x209693Bc6afc0C5328bA36FaF03C514EF312287C".to_string(),
        max_timeout_seconds: 300,
        asset: "0x036CbD53842c5426634e7929541eC2318f3dCF7e".to_string(),
        extra: Some(extra),
    }
}

// Deterministic test key (same as TypeScript tests: 0xab repeated 32 times)
const TEST_PRIVATE_KEY: &str = "0xabababababababababababababababababababababababababababababababab";

#[cfg(test)]
mod account_wallet {
    use super::*;

    #[test]
    fn creates_from_private_key() {
        let wallet = AccountWallet::from_private_key(TEST_PRIVATE_KEY).unwrap();
        // Should have a valid address
        let addr = wallet.address();
        assert!(format!("{addr}").starts_with("0x"));
        assert_eq!(format!("{addr}").len(), 42);
    }

    #[test]
    fn rejects_invalid_private_key() {
        let result = AccountWallet::from_private_key("not-a-key");
        assert!(result.is_err());
    }

    #[test]
    fn rejects_empty_private_key() {
        let result = AccountWallet::from_private_key("");
        assert!(result.is_err());
    }

    #[tokio::test]
    async fn creates_exact_payment() {
        let wallet = AccountWallet::from_private_key(TEST_PRIVATE_KEY).unwrap();
        let payment = wallet.create_payment(&test_requirements()).await.unwrap();

        assert_eq!(payment.x402_version, 1);
        assert_eq!(payment.scheme, "exact");
        assert_eq!(payment.network, "base-sepolia");

        // Verify payload structure
        assert!(payment.payload.get("signature").is_some());
        assert!(payment.payload.get("authorization").is_some());

        let auth = &payment.payload["authorization"];
        assert!(auth["from"].as_str().unwrap().starts_with("0x"));
        assert_eq!(
            auth["to"].as_str().unwrap(),
            "0x209693Bc6afc0C5328bA36FaF03C514EF312287C"
        );
        assert_eq!(auth["value"].as_str().unwrap(), "10000");
    }

    #[tokio::test]
    async fn rejects_unsupported_scheme() {
        let wallet = AccountWallet::from_private_key(TEST_PRIVATE_KEY).unwrap();
        let mut req = test_requirements();
        req.scheme = "deferred".to_string();

        let result = wallet.create_payment(&req).await;
        assert!(result.is_err());
        assert!(result.unwrap_err().to_string().contains("Unsupported"));
    }

    #[tokio::test]
    async fn rejects_missing_domain_name() {
        let wallet = AccountWallet::from_private_key(TEST_PRIVATE_KEY).unwrap();
        let mut req = test_requirements();
        req.extra = Some(HashMap::new()); // no name/version

        let result = wallet.create_payment(&req).await;
        assert!(result.is_err());
        assert!(result.unwrap_err().to_string().contains("name"));
    }
}

#[cfg(test)]
mod smart_account_wallet {
    use super::*;

    fn test_config() -> SmartAccountConfig {
        SmartAccountConfig {
            smart_account_address: "0x857b06519E91e3A54538791bDbb0E22373e36b66"
                .parse()
                .unwrap(),
            session_key_private_key: TEST_PRIVATE_KEY.to_string(),
            chain_id: 84532,
            validator_address: None, // uses default OWNABLE_VALIDATOR
        }
    }

    #[test]
    fn creates_with_default_validator() {
        let wallet = SmartAccountWallet::new(test_config());
        assert_eq!(
            format!("{}", wallet.address()),
            "0x857b06519E91e3A54538791bDbb0E22373e36b66"
        );
    }

    #[tokio::test]
    async fn creates_exact_payment() {
        let wallet = SmartAccountWallet::new(test_config());
        let payment = wallet.create_payment(&test_requirements()).await.unwrap();

        assert_eq!(payment.x402_version, 1);
        assert_eq!(payment.scheme, "exact");
        assert_eq!(payment.network, "base-sepolia");

        let auth = &payment.payload["authorization"];
        assert_eq!(
            auth["from"].as_str().unwrap().to_lowercase(),
            "0x857b06519e91e3a54538791bdbb0e22373e36b66"
        );
        assert_eq!(
            auth["to"].as_str().unwrap(),
            "0x209693Bc6afc0C5328bA36FaF03C514EF312287C"
        );
        assert_eq!(auth["value"].as_str().unwrap(), "10000");
    }

    #[tokio::test]
    async fn rejects_unsupported_scheme() {
        let wallet = SmartAccountWallet::new(test_config());
        let mut req = test_requirements();
        req.scheme = "deferred".to_string();

        let result = wallet.create_payment(&req).await;
        assert!(result.is_err());
        assert!(result.unwrap_err().to_string().contains("Unsupported"));
    }
}

File: typescript/packages/ampersend-sdk/tests/x402/wallets/smart-account/wallet.test.ts
import { SmartAccountWallet, type SmartAccountConfig } from "@/x402/wallets/smart-account/wallet.ts"
import type { Address, Hex } from "viem"
import { generatePrivateKey } from "viem/accounts"
import { afterEach, beforeEach, describe, expect, it, vi } from "vitest"

describe("SmartAccountWallet", () => {
  const baseConfig: SmartAccountConfig = {
    smartAccountAddress: "0x1111111111111111111111111111111111111111" as Address,
    sessionKeyPrivateKey: generatePrivateKey() as Hex,
  }

  it("should default validatorAddress to OWNABLE_VALIDATOR", () => {
    const wallet = new SmartAccountWallet(baseConfig)
    expect(wallet.address).toBe(baseConfig.smartAccountAddress)
  })

  it("should default coSignerValidatorAddress to COSIGNER_VALIDATOR", () => {
    const wallet = new SmartAccountWallet(baseConfig)
    expect(wallet.address).toBe(baseConfig.smartAccountAddress)
  })

  it("should use provided coSignerValidatorAddress when given", () => {
    const customValidator = "0x9999999999999999999999999999999999999999" as Address
    const wallet = new SmartAccountWallet({
      ...baseConfig,
      coSignerValidatorAddress: customValidator,
    })
    expect(wallet.address).toBe(baseConfig.smartAccountAddress)
  })

  it("should reject unsupported payment schemes", async () => {
    const wallet = new SmartAccountWallet(baseConfig)
    await expect(
      wallet.createPayment({
        protocol: "x402-v2",
        request: {
          x402Version: 2,
          resource: { url: "test", description: "test", mimeType: "application/json" },
          accepts: [
            {
              scheme: "deferred",
              network: "eip155:84532",
              amount: "1000000",
              asset: "0x036CbD53842c5426634e7929541eC2318f3dCF7e",
              payTo: "0x2222222222222222222222222222222222222222",
              maxTimeoutSeconds: 300,
              extra: {},
            },
          ],
        },
        acceptsIndex: 0,
      }),
    ).rejects.toThrow("Unsupported payment scheme")
  })

  describe("chainId derivation", () => {
    it("signs a v2 instruction; chain id is derived from CAIP-2", async () => {
      const wallet = new SmartAccountWallet(baseConfig)
      const auth = await wallet.createPayment({
        protocol: "x402-v2",
        request: {
          x402Version: 2,
          resource: { url: "https://api.example.com/x" },
          accepts: [
            {
              scheme: "exact",
              network: "eip155:10", // Optimism
              amount: "1000000",
              asset: "0x036CbD53842c5426634e7929541eC2318f3dCF7e",
              payTo: "0x2222222222222222222222222222222222222222",
              maxTimeoutSeconds: 300,
              extra: { name: "USDC", version: "2" },
            },
          ],
        },
        acceptsIndex: 0,
      })
      expect(auth.protocol).toBe("x402-v2")
    })

    it("signs a v1 instruction; chain id is derived from the network name", async () => {
      const wallet = new SmartAccountWallet(baseConfig)
      const auth = await wallet.createPayment({
        protocol: "x402-v1",
        request: {
          x402Version: 1,
          accepts: [
            {
              scheme: "exact",
              network: "base-sepolia",
              maxAmountRequired: "1000000",
              resource: "https://api.example.com/x",
              description: "Test",
              mimeType: "application/json",
              payTo: "0x2222222222222222222222222222222222222222",
              maxTimeoutSeconds: 300,
              asset: "0x036CbD53842c5426634e7929541eC2318f3dCF7e",
              extra: { name: "USDC", version: "2" },
            },
          ],
        },
        acceptsIndex: 0,
      })
      expect(auth.protocol).toBe("x402-v1")
    })

    it("throws when the chain id cannot be derived (non-EVM v2 namespace)", async () => {
      const wallet = new SmartAccountWallet(baseConfig)
      await expect(
        wallet.createPayment({
          protocol: "x402-v2",
          request: {
            x402Version: 2,
            resource: { url: "https://api.example.com/x" },
            accepts: [
              {
                scheme: "exact",
                network: "solana:mainnet" as never,
                amount: "1000000",
                asset: "So11111111111111111111111111111111111111112",
                payTo: "0x2222222222222222222222222222222222222222",
                maxTimeoutSeconds: 300,
                extra: { name: "USDC", version: "2" },
              },
            ],
          },
          acceptsIndex: 0,
        }),
      ).rejects.toThrow(/Unsupported network/)
    })

    it("throws when the v1 network name is unknown", async () => {
      const wallet = new SmartAccountWallet(baseConfig)
      await expect(
        wallet.createPayment({
          protocol: "x402-v1",
          request: {
            x402Version: 1,
            accepts: [
              {
                scheme: "exact",
                network: "not-a-real-chain",
                maxAmountRequired: "1000000",
                resource: "https://api.example.com/x",
                description: "Test",
                mimeType: "application/json",
                payTo: "0x2222222222222222222222222222222222222222",
                maxTimeoutSeconds: 300,
                asset: "0x036CbD53842c5426634e7929541eC2318f3dCF7e",
                extra: { name: "USDC", version: "2" },
              },
            ],
          },
          acceptsIndex: 0,
        }),
      ).rejects.toThrow(/Unsupported network/)
    })
  })

  describe("authorization timestamps", () => {
    const PINNED_UNIX_SECONDS = 1_700_000_000
    const CLOCK_SKEW_BACKDATE_SECONDS = 600
    const maxTimeoutSeconds = 300

    beforeEach(() => {
      vi.useFakeTimers()
      vi.setSystemTime(PINNED_UNIX_SECONDS * 1000)
    })
    afterEach(() => {
      vi.useRealTimers()
    })

    it("derives validAfter and validBefore from a single clock read", async () => {
      const wallet = new SmartAccountWallet(baseConfig)
      const auth = await wallet.createPayment({
        protocol: "x402-v1",
        request: {
          x402Version: 1,
          accepts: [
            {
              scheme: "exact",
              network: "base-sepolia",
              maxAmountRequired: "1000000",
              resource: "https://api.example.com/x",
              description: "Test",
              mimeType: "application/json",
              payTo: "0x2222222222222222222222222222222222222222",
              maxTimeoutSeconds,
              asset: "0x036CbD53842c5426634e7929541eC2318f3dCF7e",
              extra: { name: "USDC", version: "2" },
            },
          ],
        },
        acceptsIndex: 0,
      })

      if (auth.protocol !== "x402-v1") throw new Error("expected v1")
      const inner = auth.data.payload as { authorization: Record<string, string> }
      expect(inner.authorization.validAfter).toBe(String(PINNED_UNIX_SECONDS - CLOCK_SKEW_BACKDATE_SECONDS))
      expect(inner.authorization.validBefore).toBe(String(PINNED_UNIX_SECONDS + maxTimeoutSeconds))
    })
  })
})
